src/consumers/consumers.c

```diff
@@ -2,15 +2,118 @@
+ * @arg args - (consumer_args), see ~/statsd-parsers/statsd-parsers.h
+ */
 void* consume_datagram(void* args) {
     chan_t* parsed = ((consumer_args*)args)->parsed_datagrams;
+    agent_config* config = ((consumer_args*)args)->config;
+    metrics* m = ((consumer_args*)args)->metrics_wrapper;
+    init_duration_consumer(config);
+    init_counter_consumer(config);
+    init_gauge_consumer(config);
     statsd_datagram* datagram = (statsd_datagram*) malloc(sizeof(statsd_datagram));
     while(1) {
         *datagram = (statsd_datagram) { 0 };
-        chan_recv(parsed, (void *)&datagram);
-        print_out_datagram(datagram);
+        switch(chan_select(&parsed, 1, (void *)&datagram, NULL, 0, NULL)) {
+            case 0:
+                {
+                    if (strcmp(datagram->type, "ms") == 0) {
+                        process_duration(m, datagram);
+                    } else if (strcmp(datagram->type, "c") == 0) {
+                        process_counter(m, datagram);
+                    } else if (strcmp(datagram->type, "g") == 0) {
+                        process_gauge(m, datagram);
+                    }
+                }
+                break;
+            default:
+                {
+                    if (g_output_requested) {
+                        verbose_log("Output of recorded values request caught.");
+                        print_recorded_values(m, config);
+                        verbose_log("Recorded values output.");
+                        g_output_requested = 0;
+                    }
+                }
+        }
+    }
+}
+
+/**
+ * Sets flag notifying that output was requested
+ */
+void consumer_request_output() {
+    g_output_requested = 1;
+}
+
+/**
+ * Checks if given metric name is available (it isn't recorded yet)
+ * @arg m - Metrics struct (storage in which to check for name availability)
+ * @arg name - Name to be checked
+ * @return 1 on success else 0
+ */
+int check_metric_name_available(metrics* m, char* name) {
+    if (!find_counter_by_name(m, name, NULL) ||
+        !find_gauge_by_name(m, name, NULL) ||
+        !find_histogram_by_name(m, name, NULL)) {
+            return 1;
+        }
+    return 0;
+}
+
+/**
+ * Prints values of currently recorded metrics with total count, if config allows it, into file that is specified in it under "debug" key
+ * @arg m - Metrics struct (what values to print)
+ * @arg config - Config to check against
+ */
+void print_recorded_values(metrics* m, agent_config* config) {
+    if (strlen(config->debug_output_filename) == 0) return; 
+    FILE* f;
+    f = fopen(config->debug_output_filename, "w+");
+    long long int record_count = 0;
+    if (f != NULL) {
+        record_count += print_counter_metric_collection(m, f);
+        record_count += print_gauge_metric_collection(m, f);
+        record_count += print_duration_metric_collection(m, f);
     }
+    fprintf(f, "Total number of records: %llu \n", record_count);
+    fclose(f);
```

Comment:
This is a bit fragile code, fopen can return `NULL` which will break on line 112.

Reply:
Fixed by bc15823